File: 0047-permutations-ii/0047-permutations-ii.java
class Solution {
    void Func(int[] nums, List<Integer> temp, List<List<Integer>> res, boolean[] used, int start)
    {
        if(temp.size() == nums.length)
        {
            res.add(new ArrayList<>(temp));
            return;
        }
        for(int i=0; i<nums.length; i++)
        {
            if(used[i]) continue;
            if(i > 0 && nums[i] == nums[i - 1] && !used[i - 1])
                continue;
            used[i] = true;
            temp.add(nums[i]);
            Func(nums, temp, res, used, i+1);
            temp.remove(temp.size() - 1);
            used[i] = false;
        }
    }
    public List<List<Integer>> permuteUnique(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        boolean used[] = new boolean[nums.length];
        Arrays.sort(nums);
        Func(nums, new ArrayList<>(), res, used, 0);
        return res; 
    }
}

File: 0047-permutations-ii/0047-permutations-ii.cpp
class Solution {
public:
    void Func(vector<int>& nums, vector<int>& temp, vector<vector<int>>& res, vector<bool>& used, int start)
    {
        if(temp.size() == nums.size())
        {
            res.push_back(temp);
            return;
        }
        for(int i=0; i<nums.size(); i++)
        {
            if(used[i]) continue;
            if(i > 0 && nums[i] == nums[i - 1] && !used[i - 1])
                continue;
            used[i] = true;
            temp.push_back(nums[i]);
            Func(nums, temp, res, used, i+1);
            temp.pop_back();
            used[i] = false;

        }
    }
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        vector<vector<int>> res;
        vector<int> temp;
        vector<bool>used(nums.size(), false);
        sort(nums.begin(), nums.end());
        Func(nums, temp, res, used, 0);
        return res;
    }
};